### DISCUSSION TITLE
return native PostgreSQL ISO8601 format

### DISCUSSION THREAD

**Discussion by seven1240** [PostgREST/postgrest#2924] [Q&A] [answered]
### Environment

* PostgreSQL version: docker `postgres:14.0-alpine`
* PostgREST version: docker `v11.2.0`
* Operating system: `Darwin Kernel Version 22.2.0 root:xnu-8792.61.2~4/RELEASE_ARM64_T6000 arm64`

### Description of issue

PostgreSQL by default returns ISO 8601 format withouth `T` between date and time for easy reading, e.g.

```
2023-08-25 21:50:45
v.s.
2023-08-25T21:50:45
```

But PostgREST has `T` in it. Looks like it was without `T` in 2015: https://github.com/PostgREST/postgrest/issues/177

Anyway, is it possible to add a config option to choose the format?

Thanks.

**Comment by steve-chavez** [comment]
@seven1240 I always get the `T` inside the timestamp:

```sql
create table article_stars (
    article_id integer not null,
    user_id integer not null,
    created_at timestamp without time zone default now() not null,
    primary key (article_id, user_id)
);

curl localhost:3000/article_stars
[{"article_id":1,"user_id":1,"created_at":"2015-12-08T04:22:57.472738"}, 
 {"article_id":1,"user_id":2,"created_at":"2015-12-08T04:22:57.472738"}, 
 {"article_id":2,"user_id":3,"created_at":"2015-12-08T04:22:57.472738"}, 
 {"article_id":3,"user_id":2,"created_at":"2015-12-08T04:22:57.472738"}, 
 {"article_id":1,"user_id":3,"created_at":"2015-12-08T04:22:57.472738"}]
```

To remove it, you could use a [computed field](https://postgrest.org/en/stable/references/api/computed_fields.html) or [domain representation](https://postgrest.org/en/stable/references/api/domain_representations.html).

---

For domain representations, this example could help:

https://github.com/PostgREST/postgrest/blob/531a183b44b36614224fda432335cdaa356b4a0a/test/spec/fixtures/schema.sql#L3128-L3144

**Comment by steve-chavez** [comment]
Additionally you could try changing the `datestyle` for the authenticator:

```sql
alter role authenticator set datestyle to postgres;
```

**Comment by seven1240** [comment]
Hi @steve-chavez , thank you for the detailed explain. The domain way seems is little complicated.

> Additionally you could try changing the `datestyle` for the authenticator:
> 
> ```sql
> alter role authenticator set datestyle to postgres;
> ```

I tried this and no effect to me even after I restarted PostgREST. Also, `postgres` is not the right style I want. That's what I saw in psql:

```
xui=> show datestyle;
 DateStyle
-----------
 ISO, MDY
(1 row)

xui=> select now();
              now
-------------------------------
 2023-08-30 22:24:22.917144+08
(1 row)

xui=> set datestyle = postgres;
SET
xui=> select now();
                 now
-------------------------------------
 Wed Aug 30 22:24:31.516747 2023 CST
(1 row)
```


According to https://www.postgresql.org/docs/current/datatype-datetime.html :

> ISO 8601 specifies the use of uppercase letter T to separate the date and time. PostgreSQL accepts that format on input, but on output it uses a space rather than T, as shown above. This is for readability and for consistency with [RFC 3339](https://tools.ietf.org/html/rfc3339) as well as some other database systems.

So back in the original question, Looks like PostgREST in a 2015 version has not `T` and does match what we see in psql  as described in https://github.com/PostgREST/postgrest/issues/177 . What changed from then? 

I guess the actual format is still formatted by PostgREST, or the low level pg client lib?

Thanks again.

**Reply by wolfgangwalther** [reply, marked as answer]
This format is used when returning a timestamp as json:

```sql
postgres=> SELECT to_json(now());
              to_json
------------------------------------
 "2023-08-30T15:04:48.724897+00:00"
(1 row)
```

Since we use the built-in transformation to json now, all columns exposed as type timestamp will have this format. However, if you cast your columns to text first, you will receive what you want:

```sql
postgres=> SELECT to_json(now()::TEXT);
             to_json
---------------------------------
 "2023-08-30 15:04:44.512372+00"
(1 row)
```

Datestyle settings will only affect the cast to TEXT (2nd example), not the cast to json directly.

You can achieve the cast either via `VIEW`s, via computed columns (as mentioned above) or via domains (also mentioned above).
